protocols/upnp/src/behaviour.rs

```diff
@@ -0,0 +1,554 @@
+    /// Port mapping failed, we will try again.
+    Failed,
+    /// Port mapping is permanent on the Gateway.
+    Permanent,
+}
+
+/// Current state of the UPnP [`Gateway`].
+enum GatewayState {
+    Searching(BoxFuture<'static, Result<Gateway, Box<dyn std::error::Error>>>),
+    Available(Gateway),
+    GatewayNotFound,
+}
+
+/// The event produced by `Behaviour`.
+#[derive(Debug)]
+pub enum Event {
+    /// The multiaddress is reachable externally.
+    NewExternalAddr(Multiaddr),
+    /// The renewal of the multiaddress on the gateway failed.
+    ExpiredExternalAddr(Multiaddr),
+    /// The IGD gateway was not found.
+    GatewayNotFound,
+}
+
+/// A `NetworkBehaviour` for UPnP port mapping. Automatically tries to map the external port
+/// to an internal address on the gateway on a `FromSwarm::NewListenAddr`.
+pub struct Behaviour<P>
+where
+    P: Provider,
+{
+    /// Gateway config.
+    config: Config,
+    /// UPnP interface state.
+    state: GatewayState,
+
+    /// List of port mappings.
+    mappings: HashMap<Mapping, MappingState>,
+
+    /// Pending behaviour events to be emitted.
+    pending_events: VecDeque<Event>,
+
+    /// Provider.
+    provider: PhantomData<P>,
+}
+
+impl<P> Default for Behaviour<P>
+where
+    P: Provider + 'static,
+{
+    fn default() -> Self {
+        Self::new(Config::default())
+    }
+}
+
+impl<P> Behaviour<P>
+where
+    P: Provider + 'static,
+{
+    /// Builds a new `UPnP` behaviour.
+    pub fn new(config: Config) -> Self {
+        Self {
+            config,
+            state: GatewayState::Searching(P::search_gateway(config).boxed()),
+            mappings: Default::default(),
+            pending_events: VecDeque::new(),
+            provider: PhantomData,
+        }
+    }
+}
+
+impl<P> NetworkBehaviour for Behaviour<P>
+where
+    P: Provider + 'static,
+{
+    type ConnectionHandler = dummy::ConnectionHandler;
+
+    type ToSwarm = Event;
+
+    fn handle_established_inbound_connection(
+        &mut self,
+        _connection_id: ConnectionId,
+        _peer: PeerId,
+        _local_addr: &Multiaddr,
+        _remote_addr: &Multiaddr,
+    ) -> Result<libp2p_swarm::THandler<Self>, ConnectionDenied> {
+        Ok(dummy::ConnectionHandler)
+    }
+
+    fn handle_established_outbound_connection(
+        &mut self,
+        _connection_id: ConnectionId,
+        _peer: PeerId,
+        _addr: &Multiaddr,
+        _role_override: Endpoint,
+    ) -> Result<libp2p_swarm::THandler<Self>, libp2p_swarm::ConnectionDenied> {
+        Ok(dummy::ConnectionHandler)
+    }
+
+    fn on_swarm_event(&mut self, event: FromSwarm<Self::ConnectionHandler>) {
+        match event {
+            FromSwarm::NewListenAddr(NewListenAddr {
+                listener_id,
+                addr: multiaddr,
+            }) => {
+                let (addr, protocol) = match multiaddr_to_socketaddr_protocol(multiaddr.clone()) {
+                    Ok(addr_port) => addr_port,
+                    Err(_) => {
```

Comment:
```suggestion
                    Err(()) => {
```

In case the type ever changes in the future, we get a compile time error.

Reply:
Thanks Max, makes sense. Updated :)